system-addon/content-src/lib/init-store.js

```diff
@@ -76,6 +77,28 @@ const rehydrationMiddleware = store => next => action => {
   return next(action);
 };
 
+// This middleware queues up all the EARLY_QUEUED_ACTIONS until it receives
+// the first action from main. Note that NEW_TAB_INITIAL_STATE is guaranteed
+// to be sent from main
+const queueEarlyMessageMiddleware = store => next => action => {
+  if (store._receivedFromMain) {
+    next(action);
+  } else if (au.isFromMain(action)) {
+    next(action);
+    store._receivedFromMain = true;
+    if (store._queue && store._queue.length) {
```

Comment:
The `length` check shouldn't be necessary. And the `_queue` name is a bit generic especially we're just attaching to some effectively shared/global `store` object.

Reply:
Also, add a comment here saying that we're sending out any early actions /after/ processing the message from main.